File: npm/src/factory.ts
/**
 * Signal Factory Functions
 *
 * Convenience functions for creating HCT signals.
 */

import {
    HCTSignal,
    SignalType,
    Tempo,
    HoldType,
} from './types';

/** Options for creating a CUE signal */
export interface CueOptions {
    source: string;
    targets: string[];
    payload?: Record<string, unknown>;
    urgency?: number;
    tempo?: Tempo;
}

/** Create a CUE signal to trigger agent activation */
export function cue(options: CueOptions): HCTSignal {
    return {
        type: SignalType.CUE,
        source: options.source,
        targets: options.targets,
        payload: options.payload ?? {},
        performance: {
            urgency: options.urgency ?? 5,
            tempo: options.tempo ?? Tempo.MODERATO,
        },
        timestamp: new Date().toISOString(),
    };
}

/** Options for creating a FERMATA signal */
export interface FermataOptions {
    source: string;
    reason: string;
    holdType?: HoldType;
    timeout_ms?: number;
}

/** Create a FERMATA signal to hold for approval */
export function fermata(options: FermataOptions): HCTSignal {
    return {
        type: SignalType.FERMATA,
        source: options.source,
        targets: ['governance'],
        payload: { reason: options.reason },
        conditions: { hold_type: options.holdType ?? HoldType.HUMAN },
        performance: { timeout_ms: options.timeout_ms },
        timestamp: new Date().toISOString(),
    };
}

/** Options for creating an ATTACCA signal */
export interface AttaccaOptions {
    source: string;
    targets: string[];
    payload?: Record<string, unknown>;
}

/** Create an ATTACCA signal for immediate transition */
export function attacca(options: AttaccaOptions): HCTSignal {
    return {
        type: SignalType.ATTACCA,
        source: options.source,
        targets: options.targets,
        payload: options.payload ?? {},
        performance: { urgency: 10, tempo: Tempo.PRESTO },
        timestamp: new Date().toISOString(),
    };
}

/** Options for creating a VAMP signal */
export interface VampOptions {
    source: string;
    repeatUntil: string;
    qualityThreshold?: number;
    timeout_ms?: number;
}

/** Create a VAMP signal to repeat until condition met */
export function vamp(options: VampOptions): HCTSignal {
    return {
        type: SignalType.VAMP,
        source: options.source,
        targets: [options.source], // Self-loop
        payload: {},
        conditions: {
            repeat_until: options.repeatUntil,
            quality_threshold: options.qualityThreshold ?? 0.9,
        },
        performance: { timeout_ms: options.timeout_ms ?? 60000 },
        timestamp: new Date().toISOString(),
    };
}

/** Options for creating a CAESURA signal */
export interface CaesuraOptions {
    source: string;
    reason: string;
}

/** Create a CAESURA signal for full stop */
export function caesura(options: CaesuraOptions): HCTSignal {
    return {
        type: SignalType.CAESURA,
        source: options.source,
        targets: ['*'], // Broadcast
        payload: { reason: options.reason },
        performance: { urgency: 10, tempo: Tempo.PRESTO },
        timestamp: new Date().toISOString(),
    };
}

/** Options for creating a TACET signal */
export interface TacetOptions {
    source: string;
    duration_ms?: number;
}

/** Create a TACET signal to mark agent as inactive */
export function tacet(options: TacetOptions): HCTSignal {
    return {
        type: SignalType.TACET,
        source: options.source,
        targets: [],
        payload: options.duration_ms ? { duration_ms: options.duration_ms } : {},
        timestamp: new Date().toISOString(),
    };
}

/** Options for creating a DOWNBEAT signal */
export interface DownbeatOptions {
    source: string;
    syncPoint: string;
}

/** Create a DOWNBEAT signal for global synchronization */
export function downbeat(options: DownbeatOptions): HCTSignal {
    return {
        type: SignalType.DOWNBEAT,
        source: options.source,
        targets: ['*'], // Broadcast
        payload: { sync_point: options.syncPoint },
        timestamp: new Date().toISOString(),
    };
}

File: python/hct_mcp_signals/factory.py
"""
HCT Signal Factory Functions

Convenience functions for creating HCT signals.
"""

from typing import Any, Optional

from .schema import Conditions, HCTSignal, HoldType, Performance, SignalType, Tempo


def cue(
    source: str,
    targets: list[str],
    payload: Optional[dict[str, Any]] = None,
    urgency: int = 5,
    tempo: str = "moderato",
) -> HCTSignal:
    """Create a CUE signal to trigger agent activation."""
    return HCTSignal(
        type=SignalType.CUE,
        source=source,
        targets=targets,
        payload=payload or {},
        performance=Performance(urgency=urgency, tempo=Tempo(tempo)),
    )


def fermata(
    source: str, reason: str, hold_type: str = "human", timeout_ms: Optional[int] = None
) -> HCTSignal:
    """Create a FERMATA signal to hold for approval."""
    return HCTSignal(
        type=SignalType.FERMATA,
        source=source,
        targets=["governance"],
        payload={"reason": reason},
        performance=Performance(timeout_ms=timeout_ms),
        conditions=Conditions(hold_type=HoldType(hold_type)),
    )


def attacca(
    source: str, targets: list[str], payload: Optional[dict[str, Any]] = None
) -> HCTSignal:
    """Create an ATTACCA signal for immediate transition."""
    return HCTSignal(
        type=SignalType.ATTACCA,
        source=source,
        targets=targets,
        payload=payload or {},
        performance=Performance(urgency=10, tempo=Tempo.PRESTO),
    )


def vamp(
    source: str,
    repeat_until: str,
    quality_threshold: float = 0.9,
    timeout_ms: int = 60000,
) -> HCTSignal:
    """Create a VAMP signal to repeat until condition met."""
    return HCTSignal(
        type=SignalType.VAMP,
        source=source,
        targets=[source],  # Self-loop
        payload={},
        performance=Performance(timeout_ms=timeout_ms),
        conditions=Conditions(
            repeat_until=repeat_until, quality_threshold=quality_threshold
        ),
    )


def caesura(source: str, reason: str) -> HCTSignal:
    """Create a CAESURA signal for full stop."""
    return HCTSignal(
        type=SignalType.CAESURA,
        source=source,
        targets=["*"],  # Broadcast
        payload={"reason": reason},
        performance=Performance(urgency=10, tempo=Tempo.PRESTO),
    )


def tacet(source: str, duration_ms: Optional[int] = None) -> HCTSignal:
    """Create a TACET signal to mark agent as inactive."""
    return HCTSignal(
        type=SignalType.TACET,
        source=source,
        targets=[],
        payload={"duration_ms": duration_ms} if duration_ms else {},
    )


def downbeat(source: str, sync_point: str) -> HCTSignal:
    """Create a DOWNBEAT signal for global synchronization."""
    return HCTSignal(
        type=SignalType.DOWNBEAT,
        source=source,
        targets=["*"],  # Broadcast
        payload={"sync_point": sync_point},
    )
